### PR TITLE
Fix vm_compute evar sensitivity with decompose_prod

### PR DESCRIPTION
**Pull request by SkySkimmer**
Without this https://github.com/coq/coq/pull/17346 causes error in bug_6378.v
`vm_compute in (Z.div_eucl, Z.div_eucl, Z.div_eucl, Z.div_eucl, Z.div_eucl)`

### CODE REVIEW

File: pretyping/vnorm.ml [hunk 1]
```diff
@@ -33,8 +32,8 @@ let e_whd_all = Reductionops.clos_whd_flags CClosure.all
 
 let crazy_type =  mkSet
 
-let decompose_prod env t =
-  let (name,dom,codom) = destProd (whd_all env t) in
+let decompose_prod env sigma t =
+  let (name,dom,codom) = destProd (EConstr.Unsafe.to_constr (e_whd_all env sigma (EConstr.of_constr t))) in
```
Comment (SkySkimmer) [inline pretyping/vnorm.ml:36]: The Unsafe.to_constr here is not pretty but trying to do without it I was fighting the types more than seemed worth doing.

**Review by ppedrot — APPROVED** [review]
The code change is fine, but can we get a test maybe?

**Comment by silene** [comment]
It seems like something that was meant to be close to a weak normal form now requires some evar info to be so. But, the corollary is that the virtual machine can no longer complete its reduction job due to missing evar info, so the lazy reduction engine has to take over the job. That seems like a slippery slope, as this pull request is just there to paper over a future bug. Am I misunderstanding something?

**Comment by SkySkimmer** [comment]
AFAIK in vm_compute we need to reduce types at times for the reification, and we always use lazy for this.

**Comment by ppedrot** [comment]
Yes, this patch is about normalization of types, not terms. And indeed we've always used the kernel machine for (most of) type normalization.

**Comment by ppedrot** [comment]
@coqbot run full ci

**Comment by ppedrot** [comment]
There are already cases where we call the kernel reduction on arbitrary types, not necessarily containing evars. For instance, when normalizing pattern-matching for instance. There the return clause of the match is totally reduced kernel-side.

**Comment by SkySkimmer** [comment]
>when v1 is an accumulator to an evar that did not get unfolded. (If the evar had been unfolded, the VM would have taken care of reducing v1, not the kernel.)

Can that actually happen? AFAIK we only send terms with all evars unfolded to the VM, so it doesn't make sense to get back a folded evar. Only `T` could have folded evars.

**Comment by silene** [comment]
> There the return clause of the match is totally reduced kernel-side.

What do you mean? I am pretty sure that, for the following function, the return clause is fully reduced by the bytecode interpreter, not the kernel.
```coq
Require Import Arith.

Definition f n x :=
  match x return if x then (fact n - fact n) - (fact n - fact n) = 0 else True with
  | true => Nat.sub_diag (fact n - fact n)
  | false => I
  end.

Time Eval compute in fun x => f 8 x. (* the kernel cannot reduce "fact 9" *)
Time Eval vm_compute in fun x => f 10 x. (* yet the return clause is fully reduced with "fact 10" *)
```

> Can that actually happen?

That is my question. Can it happen with #17346?

**Comment by SkySkimmer** [comment]
Unless you think my reasoning is incorrect then I say it can't happen.

**Comment by ppedrot** [comment]
@silene this is true for the return clause itself, but if you have a commutative cut where the match returns a function, the VM-normalized clause P will be applied to the VM-normalized arguments, but reduced by the kernel machine. See e.g. the snippet below that creates a unreduced `tcase` and sends that to the stack-normalizing function without further reduction
https://github.com/coq/coq/blob/766c7fb8c2b9b42291882777bb5df092dd6bfe89/pretyping/vnorm.ml#L292-L310

**Comment by silene** [comment]
> Unless you think my reasoning is incorrect then I say it can't happen.

You said yourself that it does happen for `Eval vm_compute in (Z.div_eucl, ...)`. Which folded evar is it that breaks it? (The only one I can imagine is the parameter of the pair type, but I don't understand why it would be involved in `decompose_prod`.)

**Comment by SkySkimmer** [comment]
I don't remember, run it in the debugger if it really matters.

**Comment by ppedrot** [comment]
@silene are you convinced by the code path in my above post? The PR doesn't change what could already happen in master. I think we can try to only use the VM even for type normalization, but this PR is somewhat orthogonal to it. If that's fine with you I'll just merge it.

**Comment by silene** [comment]
Sure. I have no problem with this pull request in itself. I am just worried that it might be hiding a bug in #17346.

**Comment by ppedrot** [comment]
@coqbot merge now

**Pull request merged**